src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs

```diff
@@ -330,5 +333,40 @@ public class ParameterBinder
                     metadata);
             }
         }
+
+        private void RecalculateModelMetadata(
+            ParameterDescriptor parameter,
+            ModelBindingResult modelBindingResult,
+            ref ModelMetadata metadata)
+        {
+            // Attempt to recalculate ModelMetadata for top level parameters and properties using the actual
+            // model type. This would ensure validation uses a combination of top-level validation metadata
+            // as well as metadata on the actual, rather than declared, model type.
+
+            if (!modelBindingResult.IsModelSet ||
+                modelBindingResult.Model == null ||
+                !(_modelMetadataProvider is ModelMetadataProvider modelMetadataProvider))
```

Comment:
That's asserted in the calling code. Leaving as is.

Reply:
I didn't suggest an assertion, just that condition is unnecessary. Use a regular cast.

Reply:
See https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ParameterBinder.cs#L245-L259 - we don't require `_modelMetadataProvider` - `ModelMetadataProvider` here. The cast could throw.

Reply:
Got it. Something of a corner case since the app hasn't turned `AllowValidatingTopLevelNodes` off but has overridden the `IModelMetadataProvider` with something that isn't a `ModelMetadataProvider`. :ok: